pkg/kubectl/cmd/cmd.go

```diff
@@ -257,7 +262,80 @@ var (
 )
 
 func NewDefaultKubectlCommand() *cobra.Command {
-	return NewKubectlCommand(os.Stdin, os.Stdout, os.Stderr)
+	cmd := NewKubectlCommand(os.Stdin, os.Stdout, os.Stderr)
+
+	if len(os.Args) > 1 {
+		cmdPathPieces := os.Args[1:]
+
+		// traverse path in command tree from args given, attempting
+		// to determine the farthest path that allows extensibility
+		// if dealing with a binary extension.
+		currentCmd := cmd
+		canBeExtended := false
+		for _, target := range cmdPathPieces {
+			subCommandExists := false
+
+			for _, actual := range currentCmd.Commands() {
+				if actual.Name() == target {
+					currentCmd = actual
+					subCommandExists = true
+					break
+				}
+			}
+
+			if !subCommandExists {
+				if canBeExtended {
+					handleEndpointExtensions(cmdPathPieces)
+				}
+				break
+			}
+
+			// if subcommand exists, determine if it allows extensibility
+			if val, ok := currentCmd.Annotations[EndpointCanBeExtendedAnnotation]; ok {
+				canBeExtended = val == "true"
+			} else {
+				canBeExtended = false
+			}
+		}
+	}
+
+	return cmd
+}
+
+func handleEndpointExtensions(cmdArgs []string) {
+	remainingArgs := cmdArgs
+	for idx := range remainingArgs {
+		remainingArgs[idx] = strings.Replace(remainingArgs[idx], "-", "_", -1)
+	}
+
+	foundBinaryPath := ""
+
+	platformSuffix := ""
+	if runtime.GOOS == "windows" {
+		platformSuffix = ".exe"
+	}
+
+	// attempt to find binary, starting at longest possible name with given cmdArgs
+	for len(remainingArgs) > 1 {
+		path, err := exec.LookPath(fmt.Sprintf("kubectl-%s%s", strings.Join(remainingArgs, "-"), platformSuffix))
+		if err != nil || len(path) == 0 {
+			remainingArgs = remainingArgs[:len(remainingArgs)-1]
+			continue
+		}
+
+		foundBinaryPath = path
+		break
+	}
+
+	if len(foundBinaryPath) == 0 {
+		return
+	}
+
+	// invoke cmd binary relaying the current environment and args given
+	if err := syscall.Exec(foundBinaryPath, os.Args[1:], os.Environ()); err != nil {
```

Comment:
* does this preserve the current working dir of the kubectl invocation?
* should we pass `os.Args[len(remainingArgs):]`, so that `kubectl plugin foo bar baz` passes `baz` to `kubectl-plugin-foo-bar`?
* should we pass `cmdArgs[...]` rather than `os.Args[...]`?

Reply:
> does this preserve the current working dir of the kubectl invocation?

At least on a local env, I saw this to be true - `os.Getwd()` in kubectl returned the same value as `pwd` in the plugin executable when run via the execve call.

> should we pass os.Args[len(remainingArgs):], so that kubectl plugin foo bar baz passes baz to kubectl-plugin-foo-bar?

I was not sure about this since there was some discussion about passing args "as-is", but I'm not against this.